### DISCUSSION TITLE
Android Routing only specific traffic through the tunnel

### DISCUSSION THREAD

**Discussion by karnilamit** [vpnhood/VpnHood#33] [Q&A] [answered]
Hi there,

I have been going through the VpnHood code and have few queries regarding how we can have only specific traffic redirected through the VPN tunnel. I am testing the client on my local network which has a webserver running on IP address 192.168.0.190. It also has some other webservers running on 192.168.0.195 and so on. I need the Vpnhood to route only traffic destined through 192.168.0.190 through the VPN tunnel and the rest traffic should follow the usual route. How this can be achieved?

I also see that in code all the traffic is received on the Android capture interface and then it is pass through if the IP address is not in the range, instead of receiving all the traffic can we configure the Android VPN to only route specific traffic and not all traffic.

-AK

**Comment by trudyhood** [comment, marked as answer]
Please note that VpnHood is designed to circumvent firewalls, and your need to tunnel specific traffic from your local network makes no sense.

by the way, you can achieve that by calling AddRoute on android VpnService to filter it at the driver level.
VpnHood uses two methods for filtering traffics due to the limitation of VpnService AddRoute, the first method is using AddRoute and the other is by creating a local proxy server and redirecting the traffics. By default, VpnHood excludes all local traffics at driver level and doesn't let them go through the tunnel.

I don't know how you plan to use VpnHood for that, but there are many ways to achieve that via VpnHood modules or apps.
If you are going to use VpnHood to achieve that, you can set **ClientOptions.IncludeIpRanges** or **ClientOptions.ExcludeIpRanges** but these are not working at Driver Level (AddRoute). You can also set **ClientOptions.PacketCaptureExcludeIpRange** to exclude some IPs at driver level but exclude has limitation on Android and you can add just a few exclude items. I can also add **ClientOptions.PacketCaptureIncludeIpRange** for you if you think it is useful.

Let me know if you have more questions.

**Reply by karnilamit** [reply]
Thanks much for your reply. I used local LAN as an example. The specific scenario I am talking about is as below
![image](https://user-images.githubusercontent.com/87796890/129570691-3f5ad5c6-37ad-477d-a54e-37c4f683e362.png)

Here server 23.33.238.154 should be accessed through VPNHood (ultimately using Android VPN to route the traffic), however server 142.250.77.46 should be accessed directly, without routing the traffic to the VpnHood client on the android. I am using the code from master branch and saw that all the traffic hits the VpnHood client, which then makes the decision based on the IP-Range configured whether to send those packets to the VPNHood-Server or do a passthrough. I may be wrong, but that's what seemed to be the case. Also in my testing all the local traffic was also being hit in the TcpProxy class (which I assume you say will always be pass through).

I will try the options that you suggested soon and let you know.

Again, thanks much for taking time to reply.

**Reply by trudyhood** [reply]
It is a common scenario. You can do one of the following in VpnHood

Using **PacketCaptureFilter**
```c#
VpnHoodApp.Instance.UserSettings.ExcludeLocalNetwork = false;
VpnHoodApp.Instance.UserSettings.PacketCaptureIpRanges = new[] { new IpRange(IPAddress.Parse("23.33.238.154")) };
VpnHoodApp.Instance.UserSettings.PacketCaptureIpRangesFilterMode = FilterMode.Include;
```

or Using **GenericFilter**

```c#
VpnHoodApp.Instance.UserSettings.ExcludeLocalNetwork = false;
VpnHoodApp.Instance.UserSettings.IpGroupFilters = new[] { "custom" };
VpnHoodApp.Instance.UserSettings.CustomIpRanges = new[] { new IpRange(IPAddress.Parse("23.33.238.154")) };
VpnHoodApp.Instance.UserSettings.IpGroupFiltersMode = FilterMode.Include;
```

PacketCaptureFilter is much more faster and just the packet in your IpRange will be handled by VpnHoodApp. But they have some other differences:
* you can only add a few number of IpRanges to PacketCaptureFilter.
*  All DNS packets always managed by VpnHood regarding of your IpFilters
*  GenericFilter does not support Ping, so Ping traffics in GenericFilter always go through the VpnHood tunnel.

The code in this post is new, so you need to fetch the development branch before we merge it into main branch.
Let me know if you have more questions or need something else.

**Reply by karnilamit** [reply]
Thanks, much @trudyhood. Have been busy with some other stuff lately. Will have a look at the suggested settings and verify at my end soon. Will update you with my findings soon. Do you have a discord channel that I can subscribe to for having a direct discussion on VpnHood, or do you prefer the github discussion forum only?

**Reply by trudyhood** [reply]
Please post any question you have here or feel free to start a new discussion. You are welcome.
